examples/ct/ct_handlers.go

```diff
@@ -89,19 +133,105 @@ func enforceMethod(w http.ResponseWriter, r *http.Request, method string) bool {
 }
 
 // All the handlers are wrapped so they have access to the RPC client
-func wrappedAddChainHandler(rpcClient trillian.TrillianLogClient) http.HandlerFunc {
+// TODO(Martin2112): Doesn't properly handle duplicate submissions yet but the backend
+// needs this to be implemented before we can do it here
+func wrappedAddChainHandler(c CTRequestHandlers) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		if !enforceMethod(w, r, httpMethodPost) {
+			// HTTP status code was already set
 			return
 		}
 
-		_, err := parseBodyAsJSONChain(w, r)
+		addChainRequest, err := parseBodyAsJSONChain(w, r)
 
 		if err != nil {
+			// HTTP status code was already set
 			return
 		}
 
-		http.Error(w, http.StatusText(http.StatusNotImplemented), http.StatusNotImplemented)
+		// We already checked that the chain is not empty so can move on to verification
+		validPath := verifyAddChain(addChainRequest, w, *c.trustedRoots, false)
+
+		if validPath == nil {
+			// Chain rejected by verify. HTTP status code was already set
+			return
+		}
+
+		// Build up the SCT and MerkleTreeLeaf. The SCT will be returned to the client and
+		// the leaf will become part of the data sent to the backend.
+		merkleTreeLeaf, sct, err := SignV1SCTForCertificate(c.logKeyManager, validPath[0], c.timeSource.Now())
+
+		if err != nil {
+			glog.Warningf("Failed to create / serialize SCT or Merkle leaf: %v %v", sct, err)
+			sendHttpError(w, http.StatusInternalServerError, err)
+			return
+		}
+
+		// Inputs validated, pass the request on to the back end after hashing and serializing
+		// the data for the request
+		var leafBuffer bytes.Buffer
+		if err := WriteMerkleTreeLeaf(&leafBuffer, merkleTreeLeaf); err != nil {
+			glog.Warningf("Failed to serialize merkle leaf: %v", err)
+			sendHttpError(w, http.StatusInternalServerError, err)
+			return
+		}
+
+		var logEntryBuffer bytes.Buffer
+		logEntry := NewCTLogEntry(merkleTreeLeaf, validPath)
+		if err := logEntry.Serialize(&logEntryBuffer); err != nil {
+			glog.Warningf("Failed to serialize log entry: %v", err)
+			sendHttpError(w, http.StatusInternalServerError, err)
+			return
+		}
+
+		leafHash := sha256.Sum256(leafBuffer.Bytes())
```

Comment:
This is the leaf input to the tree hasher, the prefix is applied by the hasher. It's possible that the CT specific and generic parts don't quite connect up, which should become apparent fairly soon.